### ISSUE TITLE
automatic ResourceBundle loading?

### ISSUE THREAD

**Issue by simonsilvalauinger** [sialcasa/mvvmFX#207]
Hello everyone,
is it possible to automatically load a ResourceBundle for the view?
I mean that I have for example a ResourceBundle ExampleView.properties and it will be loaded when I load a ViewTuple for ExampleView.class with the FluentViewLoader.

Regards,
Simon

**Comment by manuel-mauky** [comment]
Hi Simon,
at the moment your use case isn't possible out of the box. You have to pass the resourceBundle as argument to the FluentViewLoader in your java code. 

However we had already thought about loading a resourceBundle by naming conventions for each view. But before we can implement such a feature there are some things to think about before: 

If you have a global application wide resourceBundle you can inject this resourceBundle into all sub-views of your application. But if we would now load an additional resourceBundle for the sub-view, how would you then access the global resourceBundle? You would only get the View-specific resourceBundle injected but not the global one. 

Think of the following example:
`application.properties`

`ExampleView.class`
`ExampleView.properties`

`SubView.class`
`SubView.properties`

From the point of view of the SubView there are three resourceBundles that might be interesting but which one should we inject into the SubView? And how does the SubView then get the other ones? 

Maybe we could merge all resourceBundles of all parents for a given View? This would be a great feature in my oppinion.

Anyway, thanks for your request. We will think about this for a future release. If you have additional comments on how you think this _should_ work we appreciate your feedback :-)

**Comment by simonsilvalauinger** [comment]
OK, this feature is more complex than I thought.
Maybe it should be possible to select a resource bundle loading strategy at a global point for example at the MvvmFX class. I could imagine somthing like:

``` java
public enum ResourceLoadingStrategy {
   GLOBAL,
   MERGING,
   PER_VIEW
}
```

then using it like `MvvmFX.setResourceLoadingStrategy(PER_VIEW)` to select a global strategy or something like `FluentViewLoader.{fxml|java}View(MyViewName.class).resourceBundleLoadingStrategy(PER_VIEW)` to select a strategy for a specific view.

**Comment by manuel-mauky** [comment]
I was thinking about this feature request a bit and the more I think about it the more I like it :-).  And I think this would be possible even without new configuration. 

This would be possible but maybe we can implement this even without new configuration needed. 
When we implement this feature we should take care for two questions:
1. Does the new handling of ResourceBundle break existing applications that update to a new mvvmfx verison?
2. Is there a reasonable use case that was possible before that wouldn't be possible anymore with  a new resourceBundle handling strategy?

I'm thinking of the following strategy:
1. load the global resourceBundle (the one that was provided to the FluentViewLoader)
2. load a resourceBundle via naming conventions for the root View and merge it. Existing key-value pairs are overwritten
3. load a resourceBundle via naming conventions for all sub views and merge it. Again existing key-value pairs are overwritten .
4. repeat 3. recursively until all views are loaded. 

With this strategy existing applications won't break because they typically won't have resourcebundle files matching the naming conventions for the existing Views. And I don't think that there is a use case that wouldn't be possible anymore.

I'm asking myself: Is there any use case that I'm overlooking where a strategy of `GLOBAL` would be needed that can't be done with `MERGING`? Or is there a use case where `PER_VIEW` would be needed? Because the resourcebundle with the same name like the View would have the highest priority anyway.

**Comment by manuel-mauky** [comment]
There is another issue we should consider when we implement this feature: 
Typically you will load the application-wide/global resourceBundle with the FluentViewLoader from the Application class. But if you load another view in your app, for example when you open a second window/dialog, again you use the FluentViewLoader. 
To have access to the global resourceBundle in the dialog you would need to pass the resourceBundle to the FluentViewLoader again. This is anoying. 
Instead we could provide a method to define a global resourceBundle:

``` java
// from the application class
MvvmFX.setGlobalResourceBundle(resourceBundle);
```

This global resource bundle would then be merged with other resourceBundles like described above.

**Comment by manuel-mauky** [comment]
After some investigation I found some restrictions of using a naming convention for loading resourceBundles:
1. When loading a View directly with the FluentViewLoader we could provide a naming convention based loading of a ResourceBundle for this view. This is an alternative to the `resourceBundle` method of the FluentViewLoader.
2. However, we can't load resourceBundles for included subviews via naming convention. When a `ExampleView.fxml` has an element `<fx:include source="SubView.fxml"/>` we can't load a resourceBundle `SubView.properties`. 

The reason is the way of how mvvmFX is loading Views. We are only wrapping the FXMLLoader from JavaFX. And the FXMLLoader doesn't provide the possibility to execute own code when an included fxml file is loaded. We have an issue #233 where we discuss a refactoring of this behaviour for a future release. 

If we only support loading of resourceBundles for the directly loaded View and not for included subviews we would introduce a new source of errors and misunderstandings. 
A developer would ask "Why doesn't the resourceBundle get's loaded? It has the same name as the fxml file?" 

For this reason we will postpone this feature until #233 is fixed. We still think it's a good idea but at the moment we can't implement it completely. 
However, I've created a new issue for the global resource bundle idea. This is possible with the current architecture: #234

**Comment by manuel-mauky** [comment]
A possible workaround is to explicitly define the resource bundle when including a View via `fx:include`. Of cause this isn't as pretty as an implicit loading via naming convention but until we provide naming convention support this may be a suitable solution:

Given the following structure:
- MainView.fxml
- MainView.properties
- SubView.fxml
- SubView.properties

in MainView.fxml you can write:

``` xml
<fx:include source="SubView.fxml" resources="SubView" />
```

However there is a bug in oracles FXMLLoader. We try to provide a workaround with issue #235
